### PR TITLE
Validate my snake

### PR DESCRIPTION
**Pull request by codeallthethingz**
For issue - https://github.com/battlesnakeio/roadmap/issues/125

* request/response objects
* validate code and tests

Slight deviation from originally spec'd
- Moved the score into sub objects for each api check with number of passed / failed checks
- Added the status code returned from the snake server call.

### CODE REVIEW

**Comment by brandonb927** [comment]
I think there's some issues with your Git config as your GH user isn't used to sign 2 commits on this PR.

File: api/server_test.go [hunk 1]
```diff
@@ -158,6 +163,36 @@ func TestGetFrames(t *testing.T) {
 	require.Equal(t, http.StatusOK, rr.Code)
 }
 
+func TestValidateMySnake(t *testing.T) {
+	s, _ := createAPIServer()
+
+	req, _ := http.NewRequest("GET", "/validateMySnake?url=dsnek.heroku.com", nil)
```
Comment (brandonb927) [inline api/server_test.go]: Is `url=dsnek.heroku.com` a valid URL though? I think we would want to require protocol at least

Comment (codeallthethingz) [reply api/server_test.go]: No, but the urls get sanitized later on.  There is only a check for existence at the controller level.

**Comment by dlsteuer** [comment]
This is just my ocd kicking in, but I'm not a fan of validateMySnake, would prefer just ValidateSnake, but I can be convinced to leave the my in there

File: controller/controller.go [hunk 2]
```diff
@@ -11,10 +11,12 @@ import (
 	"github.com/battlesnakeio/engine/controller/pb"
 	"github.com/battlesnakeio/engine/rules"
 	"github.com/battlesnakeio/engine/version"
+	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"strconv"
```
Comment (dlsteuer) [inline controller/controller.go]: This should go above 3rd party libs, with the other std imports above

File: rules/validate.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		}
+	}()
+	statusCode := response.StatusCode
+	finish := time.Now().UnixNano()
+	time := int32((finish - start) / int64(time.Millisecond))
+	contents, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", statusCode, 0, err
+	}
+	raw := string(contents[:])
```
Comment (dlsteuer) [inline rules/validate.go]: you can remove the `[:]`

**Comment by dlsteuer** [comment]
This is because it looks like gometalinter introduced a new linter called gosec, so to fix this on the lines in server.go, change it from `// nolint: gas` to `// nolint: gas, gosec`, and on the 3rd one add `// nolint: gosec`

**Comment by codeallthethingz** [comment]
@dlsteuer Ready for merge.

**Comment by dlsteuer** [comment]
Thanks looks good

**Pull request merged**